src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -1202,6 +1209,95 @@ DatabaseShardImp::findAcquireIndex(
     return boost::none;
 }
 
+void
+DatabaseShardImp::scheduleFinalizeShard(std::uint32_t shardIndex)
+{
+    taskQueue_->addTask([this, shardIndex]
+    {
+        auto retry = [this, shardIndex]
+        {
```

Comment:
Should probably make sure we aren't stopping here before setting the timer.
eg. 
```
        if (isStopping())
            return;
```

Reply:
Done

Reply:
If we return early because `isStopping()` returns true, should I also log and call `removeFailedShard()` ?

Reply:
I believe if we return at this point, on the next server startup the shard will be finalized so there shouldn't be a need to log and remove. If that isn't the case, then you log and remove.

I forgot to add the stopping check in `finalizeWithRefHash` with my branch. It should come before the lock.

Reply:
I looked into it and it seems that on restart, the aborted shard will catch at line 163 in the `DatabaseShardImp::init()` function. This will call `finalizeShard`, which will proceed with the validation without first getting the reference hash. So I guess we can either remove and log before stopping, or we'd need some method of recognizing that the shard needs to get its reference hash before we call finalize on restart. Which method would you recommend?

Reply:
True, we'd skipped the ref hash check. Lets remove and log.